### ISSUE TITLE
Duplicate, repeating log entries each time Klipper/HH is restarted

### ISSUE THREAD

**Issue by ningpj** [moggieuk/Happy-Hare#534]
Everytime I restart Klipper/HH, I end up with duplicate log entries in mmu.log.  I recall a discussion about this some time ago and cant recall if the issue was fixed or not.  Restarting the pi fixes the issue.

HH 2.7.3
Fluidd 1.30.5

**Clean boot**
```
3:29:46 MMU: Happy Hare v2.7.3 running Tradrack v1.0e with 12 gates (OPERATIONAL)
         Servo in MOVE position, Encoder reads 0.0mm
         Print state is INITIALIZED. Selector is HOMED. Tool T0 selected on gate #0
         Gear stepper at 100% current and is not synced to extruder
         
         For details on TTG and EndlessSpool groups add 'DETAIL=1', for configuration add 'SHOWCONFIG=1'
         
         Gates: | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11|
         Tools: |T0 |T1 |T2 |T3 |T4 |T5 |T6 |T7 |T8 |T9 |T10|T11|
         Avail: | B | B | B | B | B | B | B | B |   |   |   |   |
         Selct: | * |-------------------------------------------- T0
         
         [T0] > (g) .. En ......... [Ex .. (t) .. Nz] UNLOADED 0.0mm (e:0.0mm)
```
**Klipper restart in fluidd**
```
03:30:55 
         
         
         MMU Startup -----------------------------------------------
         
03:30:55 
         
         
         MMU Startup -----------------------------------------------
         
03:30:59 > __MMU_BOOTUP
03:30:59 > __MMU_BOOTUP
03:30:59 (\_/)
         ( *,*)
         (")_(") Happy Hare v2.7.3 Ready...
03:30:59 (\_/)
         ( *,*)
         (")_(") Happy Hare v2.7.3 Ready...
03:30:59 Gates: | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11|
         Tools: |T0 |T1 |T2 |T3 |T4 |T5 |T6 |T7 |T8 |T9 |T10|T11|
         Avail: | B | B | B | B | B | B | B | B |   |   |   |   |
         Selct: | * |-------------------------------------------- T0
03:30:59 Gates: | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10| 11|
         Tools: |T0 |T1 |T2 |T3 |T4 |T5 |T6 |T7 |T8 |T9 |T10|T11|
         Avail: | B | B | B | B | B | B | B | B |   |   |   |   |
         Selct: | * |-------------------------------------------- T0
03:30:59 [T0] > (g) .. En ......... [Ex .. (t) .. Nz] UNLOADED 0.0mm (e:0.0mm)
03:30:59 [T0] > (g) .. En ......... [Ex .. (t) .. Nz] UNLOADED 0.0mm (e:0.0mm)
```

**Comment by ningpj** [comment]
After a few restarts to tune EREC

```
15:36:55 > CLEAR_PAUSE
15:36:55 > CLEAR_PAUSE
15:36:55 > CLEAR_PAUSE
15:36:55 > CLEAR_PAUSE
15:36:55 > CLEAR_PAUSE
15:36:55 > CLEAR_PAUSE
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 > MMU_PRINT_END STATE=complete AUTOMATIC=1
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:57 Restoring MMU gear stepper run current to 100% configured
15:36:58 Unsynced MMU from extruder
15:36:58 Unsynced MMU from extruder
15:36:58 Unsynced MMU from extruder
```

Ill have a look to see if I can spot whats going on as im sure you fixed this previousally

**Comment by ningpj** [comment]
I think its as simple as adding a simple check to see if a handler has already been attached
```
class MmuLogger:
    def __init__(self, logfile_path):
        name = os.path.splitext(os.path.basename(logfile_path))[0]
        handler = logging.handlers.TimedRotatingFileHandler(logfile_path, when='midnight', backupCount=3)
        handler.setFormatter(MultiLineFormatter('%(asctime)s %(message)s', datefmt='%H:%M:%S'))

        self.queue_listener = QueueListener(handler)
        queue_handler = QueueHandler(self.queue_listener.bg_queue)

        self.logger = logging.getLogger(name)
        self.logger.setLevel(logging.INFO)
        if not len(self.logger.handlers):
            self.logger.addHandler(queue_handler)
        self.logger.propagate = False

        # Ensure we shutdown on exit
        atexit.register(self.shutdown)

    def log(self, message):
        self.logger.info(message)

    def shutdown(self):
        self.logger.info("Shutting down the MMU logger")
        if self.queue_listener is not None:
            self.queue_listener.stop()
```

**Comment by ningpj** [comment]
tested and fixed in 3.0.1 - fluidd and mainsail UI's

**Comment by ningpj** [comment]
tested and fixed in 3.0.1 - fluidd and mainsail UI's
